cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -257,12 +264,507 @@ void all_neighbors_graph(raft::resources const& res,
+
+#pragma omp parallel sections num_threads(3)
+    {
+#pragma omp section
+      {
+        ssize_t bytes_read = pread(graph_fd, graph_buffer.data_handle(), graph_bytes, graph_offset);
+        ASSERT(bytes_read == static_cast<ssize_t>(graph_bytes),
+               "Failed to read graph data: expected %zu, got %zd",
+               graph_bytes,
+               bytes_read);
+      }
+#pragma omp section
+      {
+        ssize_t bytes_read =
+          pread(dataset_fd, dataset_buffer.data_handle(), dataset_bytes, dataset_offset);
+        ASSERT(bytes_read == static_cast<ssize_t>(dataset_bytes),
+               "Failed to read dataset data: expected %zu, got %zd",
+               dataset_bytes,
+               bytes_read);
+      }
+#pragma omp section
+      {
+        ssize_t bytes_read = pread(label_fd, label_buffer.data_handle(), label_bytes, label_offset);
+        ASSERT(bytes_read == static_cast<ssize_t>(label_bytes),
+               "Failed to read label data: expected %zu, got %zd",
+               label_bytes,
+               bytes_read);
+      }
+    }
+
+    // Log first few values from first batch for debugging
+    if (start_row == 0 && rows_to_read > 0) {
+      RAFT_LOG_DEBUG("First graph row: [%u, %u, %u, ...]",
+                     static_cast<unsigned int>(graph_buffer(0, 0)),
+                     graph_degree_int > 1 ? static_cast<unsigned int>(graph_buffer(0, 1)) : 0,
+                     graph_degree_int > 2 ? static_cast<unsigned int>(graph_buffer(0, 2)) : 0);
+      RAFT_LOG_DEBUG("First dataset row: [%f, %f, %f, ...]",
+                     static_cast<float>(dataset_buffer(0, 0)),
+                     dim > 1 ? static_cast<float>(dataset_buffer(0, 1)) : 0.0f,
+                     dim > 2 ? static_cast<float>(dataset_buffer(0, 2)) : 0.0f);
+      RAFT_LOG_DEBUG("First labels: [%u, %u, %u, ...]",
+                     static_cast<unsigned int>(label_buffer(0)),
+                     rows_to_read > 1 ? static_cast<unsigned int>(label_buffer(1)) : 0,
+                     rows_to_read > 2 ? static_cast<unsigned int>(label_buffer(2)) : 0);
+    }
+  };
+
+  for (int64_t batch_start = 0; batch_start < n_rows; batch_start += batch_size) {
+    const int64_t current_batch_size = std::min<int64_t>(batch_size, n_rows - batch_start);
+
+    RAFT_LOG_DEBUG("Reading batch: start=%ld, size=%ld (batch_size=%zu)",
+                   batch_start,
+                   current_batch_size,
+                   batch_size);
+    read_batch(batch_start, current_batch_size);
+
+    for (int64_t batch_idx = 0; batch_idx < current_batch_size; batch_idx++) {
+      const int64_t i = batch_start + batch_idx;
+
+      os.write(reinterpret_cast<char*>(&graph_degree_int), sizeof(int));
+
+      const IdxT* graph_row = &graph_buffer(batch_idx, 0);
+      os.write(reinterpret_cast<const char*>(graph_row), sizeof(IdxT) * graph_degree_int);
+
+      if (odd_graph_degree) {
+        assert(odd_graph_degree == appr_algo->maxM0_ - graph_degree_int);
```

Comment:
This assert is only active in debug mode and fails to compile due to `-Werror` 
> comparison of integer expressions of different signedness: 'int' and 'size_t'

Pleace replace it with `RAFT_EXPECTS`